### PR TITLE
Adding support for OpenBSD/NetBSD

### PR DESCRIPTION
**Pull request by rwsalie**

### CODE REVIEW

**Comment by rwsalie** [comment]
Some errors from macro, config.py  ( adding include dirs like /usr/X11R6/lib)
other things like realpath which give an empty string for inexisting file.  lock copy file

File: contrib/bobtoolz/StdAfx.h [hunk 1]
```diff
@@ -35,7 +35,7 @@
 
 #include "time.h"
 
-#if defined( __linux__ ) || defined( __FreeBSD__ ) || defined( __APPLE__ )
+#if defined( __linux__ ) || defined( __FreeBSD__ ) || defined( __APPLE__ ) || defined( __OpenBSD__ ) || defined( __NetBSD__ )
```
Comment (TTimo) [inline contrib/bobtoolz/StdAfx.h]: I'd rather we set a radiant-only define for this than carrying/changing this long line everywhere.

File: libs/str.h [hunk 2]
```diff
@@ -47,6 +47,11 @@
   #endif
 #endif
 
+#ifdef NULL
+#undef NULL
+#define NULL 0
+#endif
```
Comment (TTimo) [inline libs/str.h]: hu? at the very least make this BSD only

File: radiant/main.cpp [hunk 3]
```diff
@@ -149,6 +149,7 @@ char *loki_gethomedir( void ){
 
 /* Must be called BEFORE loki_initialize */
 void loki_setgamename( const char *n ){
+  std::cout << n << std::endl;
```
Comment (TTimo) [inline radiant/main.cpp]: please none of that iostream stuff

File: radiant/missing.cpp [hunk 4]
```diff
@@ -56,7 +56,9 @@ bool radCopyFile( const char *lpExistingFileName, const char *lpNewFileName, boo
 	char realsrc[PATH_MAX], realdest[PATH_MAX];
 
 	realpath( lpExistingFileName, realsrc );
-	realpath( lpNewFileName, realdest );
+#ifndef __OpenBSD__
+  realpath( lpNewFileName, realdest );
+#endif
```
Comment (TTimo) [inline radiant/missing.cpp]: "realpath return empty string for an inexisting file" - doesn't sound like a BSD specific problem, so the fix should apply everywhere?

**Review by TTimo — CHANGES_REQUESTED** [review]

File: contrib/bobtoolz/StdAfx.h (hunk 1 above)
Comment (kungfooman) [reply contrib/bobtoolz/StdAfx.h]: Something like:
```c
#if defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__)
# define __BSD__
#endif
```
to shorten it?

File: libs/str.h (hunk 2 above)
Comment (rwsalie) [reply libs/str.h]: Okay i'll do that because on Clang 13.0.0 NULL is define like ((void*)0)

File: radiant/main.cpp (hunk 3 above)
Comment (rwsalie) [reply radiant/main.cpp]: sorry it was for checking stuff TT

File: radiant/missing.cpp (hunk 4 above)
Comment (rwsalie) [reply radiant/missing.cpp]: is not all BSDs specific problem only for OpenBSD ^^.
for example on NetBSD i don't get it

File: contrib/bobtoolz/StdAfx.h (hunk 1 above)
Comment (rwsalie) [reply contrib/bobtoolz/StdAfx.h]: yeah a macro like __BSD__ will help... because adding/update all defines.. take a lot of time...

**Comment by rwsalie** [comment]
Should replace  `defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__)` to  `defined(__BSD__)` ?

**Comment by rwsalie** [comment]
where should defined it on /libs/synapses.h?

**Comment by kungfooman** [comment]
This entire repo is just dead/unmaintained, right?

**Comment by rwsalie** [comment]
I don't know if I should get it back my branch alive... but i didn't get any response where i could setup a macro as __BSD__ inside GtkRadiant. (I would probably rewrite my ports, more correctly)

So if i can get any response for how i should replace __FreeBSD__ || __NetBSD__ || __OpenBSD__ would be helpful !

**Comment by TTimo** [comment]
Having one macro for all the BSD flavors sounds fine. There were other concerns raised while reviewing this pull request that are arguably more important. Please update the pull request until all concerns are adressed and we'll merge it..

**Comment by rwsalie** [comment]
TTimo can a use `__unix__` macros or you prefer keeping`__BSD__` ?

**Comment by rwsalie** [comment]
> Also as you can see Fix SIGTRAP/SIGILL* commit, we need to change char* into void against ROP on OpenBSD or undefined behaviours on other unix systems

**Comment by TTimo** [comment]
Seems like a `__BSD__` macro would be appropriate in most cases.

**Comment by rwsalie** [comment]
__BSD__ macro ready !

File: config.py [hunk 5]
```diff
@@ -259,7 +267,7 @@ def SetupEnvironment( self, env, config, useGtk = False, useGtkGL = False, useJP
         env.Append( CXXFLAGS = baseflags + [ '-fpermissive', '-fvisibility-inlines-hidden' ] )
         env.Append( CPPPATH = [ 'include', 'libs' ] )
         env.Append( CPPDEFINES = [ 'Q_NO_STLPORT' ] )
-        if ( config == 'debug' ):
+        if ( config != 'debug' ):
```
Comment (TTimo) [inline config.py]: that looks wrong

File: radiant/main.cpp [hunk 6]
```diff
@@ -186,22 +186,22 @@ int loki_getmountpoint( const char *device, char *mntpt, int max_size ){
...
-		home = const_cast<char*>(".");
-	}
-
-	strcpy( temppath, argv0 );
-
-	/* Now canonicalize it to a full pathname for the data path */
-	if ( realpath( temppath, datapath ) ) {
-		/* There should always be '/' in the path, cut after so we end our directories with a slash */
-		*( strrchr( datapath, '/' ) + 1 ) = '\0';
-	}
```
Comment (TTimo) [inline radiant/main.cpp]: this section probably needs to be rebased and reassessed, I pulled part of the fixes in already to fix the release config crash

File: radiant/missing.cpp (hunk 4 above)
Comment (TTimo) [reply radiant/missing.cpp]: I made a change to realpath usage that should avoid needing a platform define, please check and update other usage as needed

**Review by TTimo — CHANGES_REQUESTED** [review]

**Comment by rwsalie** [comment]
condition for avoiding issues from realpath workings well !

**Comment by rwsalie** [comment]
up.

**Comment by TTimo** [comment]
Sorry this is still on my list to look at and integrate, I've just been too busy on other things to get to it.

File: config.py [hunk 7]
```diff
@@ -236,7 +236,13 @@ def SetupEnvironment( self, env, config, useGtk = False, useGtkGL = False, useJP
         env.ParseConfig( 'pkg-config --libs libxml-2.0' )
         #Need to strip on xml2-config output. It has a stray \n and that completely screws up scons calling g++
         baseflags = [ '-pipe', '-Wall', '-fmessage-length=0', '-fvisibility=hidden', xml2.strip().split( ' ' ) ]
+        if ( platform.system() == "OpenBSD" ):
+            baseflags = [ '-I/usr/X11R6/include', '-I/usr/local/include', '-pipe', '-Wall', '-fmessage-length=0', '-fvisibility=hidden', xml2.strip().split( ' ' ) ]
```
Comment (TTimo) [inline config.py:240]: `/usr/local/include` ? that doesn't seem right

**Comment by TTimo** [comment]
I rebased this and pushed it to https://github.com/TTimo/GtkRadiant/tree/bsd - I'm going to review it and clean it up some more, then merge it in. I will probably squash it all together into a single commit.

**Comment by TTimo** [comment]
This is now merged : https://github.com/TTimo/GtkRadiant/commit/b5dfbd485f312a09054245a1bb1f5716b8e5e78c
Thank you for your patience with this!

**Pull request closed without merging**